pkg/upgrade/upgrade.go

```diff
@@ -0,0 +1,151 @@
+package upgrade
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/go-logr/logr"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/stolostron/hypershift-addon-operator/pkg/util"
+)
+
+type UpgradeController struct {
+	SpokeClient    client.Client
+	Log            logr.Logger
+	AddonName      string
+	AddonNamespace string
+}
+
+func (c *UpgradeController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	c.Log.Info(fmt.Sprintf("Reconciling hypershift update images configmap %s", req))
+	defer c.Log.Info(fmt.Sprintf("Done hypershift update images configmap %s", req))
+
+	upgradeRequired, err := c.upgradeImageCheck()
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if upgradeRequired {
+		c.Log.Info("image changes detected, restarting addon to upgrade the HyperShift operator")
+		os.Exit(0)
```

Comment:
Is this the most graceful way for us to solve this issue? It seems a bit hacky. There is nothing else we can do like call some k8s API etc. Not a big deal if this is the best way just a bit awkward to see.

Reply:
The other way I explored is to call the function to install the HyperShift install, which is in the `agent` package. It requires a bit of refactoring to pull that code out to a separate package to avoid a cyclic dependency. This is simplier but sure but wanted to hear if the team has big concerns first.

Reply:
Since it only runs once, can you call

What about making this call from the upgrade controller:
```
	o.Log = o.Log.WithName("agent-reconciler")
	aCtrl.plugInOption(o)

	// retry 3 times, in case something wrong with creating the hypershift install job
	if err := aCtrl.runHypershiftCmdWithRetires(ctx, 3, time.Second*10, aCtrl.runHypershiftInstall); err != nil {
		log.Error(err, "failed to install hypershift Operator")
		return err
	}
```
And remove it from the `runControllerManager`

Reply:
Refactored to move all the install/upgrade/override code to the install package. The controller no longer needs to perform os.Exit, but upgrades the HS operator instead.